ximtool/ximtool.man
.\" @(#)ximtool.1 1.1 12-Aug-2001 MJF
.TH XIMTOOL 1 "12 Aug 2001" "X11IRAF Project"
.SH NAME
ximtool \- interactive image display program for the X Window System
.sp 0.5
.SH SYNOPSIS
.B "ximtool" [\-\fItoolkitoption\fP ...] [ \fIoptions\fP ...] [\fIimagename\fP]
.sp 0.5
.SH OPTIONS
.TP 5
.B "-basePixel \fIN\fP"       
The base colormap cell used by the colormap.  This essentially allows you
to reserve \fIbasePixel\fP colors in the global colormap for other applications.
The default is 64, if changed you'll need to also specify the
\fI-cmapInitialize\fP option or resource.
.TP 5
.B "-cmap1 \fIfile\fP"          
User colormap 1.  This flag allows you to specify a colormap to be made
available at task startup.
.TP 5
.B "-cmap2 \fIfile\fP"          
User colormap 2.  This flag allows you to specify a second colormap to be
made available at task startup.
.TP 5
.B "-cmapDir1 \fIdir\fP"        
User colormap directory 1.  Specifies a directory to be searched for colormaps.
.TP 5
.B "-cmapDir2 \fIdir\fP"        
User colormap directory 2.  Specifies a directory to be searched for colormaps.
By default this points to the system directory /usr/local/lib/imtoolcmap, 
allowing a set of site default colormaps to be defined here.
.TP 5
.B "-cmapInitialize \fIbool\fP"       
Initialize the ximtool colormap at startup.  When setting the \fIbasePixel\fP
option or resource this is required in order to force the Gterm widget to 
update its global colormap resource in the X server.  The default is
\fIfalse\fP.
.TP 5
.B "-cmapName \fIname\fP"       
Name used for private colormap.  The default for all IRAF imaging
applications is \fIimage\fP.  Gterm widget based imaging applications
which have the same value of cmapName will share the same colormap,
minimizing colormap flashing and allowing multiple applications to be
run at the same time.
.TP 5
.B "-config \fIN\fP"          
Initial frame buffer configuration number.  The default value is 1, indicating
a 512x512 frame buffer with 2 frames.  See below for information on the frame
buffers.
.TP 5
.B "-defgui"                
Print the default GUI to the stdout.  The GUI is a Tcl program that may be
customized by the user and reloaded using the \fI-gui\fP option or
the \fIgui\fP resource parameter.
.TP 5
.B "-displayPanner \fIbool\fP"
Display panner marker window at startup.  If set, a panner window showing
the full frame buffer will appear in the upper-right side of the main display
window.
.TP 5
.B "-displayCoords \fIbool\fP"
Display WCS coordinate marker window at startup.  If set, a coordinate
readout text marker showing will appear in the lower-right side of the main
display window.
.TP 5
.B "-fifo \fIpipe\fP"           
Specifies the name of the fifo pipe to be used, the \fIi\fP 
and \fIo\fP suffixes will be added automatically.  The default pipe names
will be /dev/imt1i (input pipe) and /dev/imt1o (output pipe).  
.TP 5
.B "-fifo_only"             
If set, only fifo pipes will be used for communication with a client program,
sockets will be disabled.
.TP 5
.B "-gui \fIfile\fP"            
Specifies the GUI file to be used.
.TP 5
.B "-help"                  
Print a summary of command line options to the screen.
.TP 5
.B "-imtoolrc \fIfile\fP"       
Specifies the frame buffer configuration file to be used.  See below for 
information on frame buffers.
.TP 5
.B "-inet_only"
If set, only inet sockets will be used for communication with a client program,
fifo pipes and unix sockets will be disabled.
.TP 5
.B "-invert"                
Start XImtool using inverted colormaps.  When set, a "normalized" display
will always be the inverse of the selected colormap.
.TP 5
.B "-ismdev \fIdev\fP"                
Specifies the plug-in ISM connection socket.  This should be a unix domain
socket of the form "\fI/tmp/.ISM%d\fP", where the \fI%d\fP will be replaced
by the user id.  Once an ISM has connected this port is freed to accept
other connections.
.TP 5
.B "-maxColors \fIN\fP"       
Specify the max number of colors to be used for the display.
.TP 5
.B "-memModel \fItype\fP"       
Determines how ximtool uses memory in the ximtool client and the X server.  
The options are \fIfast\fP, \fIbeNiceToServer\fP, and \fIsmall\fP.  The 
default is \fIfast\fP, which uses server pixmaps to make frame blink fast.
This is recommended unless server memory is very limited.  Note that even in
fast mode, the server pixmap is only the size of the display window, so memory
usage is reasonable even if the frame buffer is very large.
.TP 5
.B "-nframes \fIN\fP"         
Specifies the number of frame buffers to configure at startup.  By default
there will be 2 frames available, a maximum of 4 frames are allowed.
.TP 5
.B "-port \fIN\fP"            
Specifies the port number to use when connecting through an inet socket.
.TP 5
.B "-port_only"
Same as \fI-inet_only\fP option.  If set, only inet sockets will be used for
communication with a client program.
.TP 5
.B "-printConfig \fIname\fP"    
Specifies the printer configuration file to use.  By default this will be
/usr/local/lib/ximprint.cfg.  See below for more information on configuring
output devices.
.TP 5
.B "-tile"                  
The default display mode is to view one frame at a time. In tile frames mode,
2 or 4 frames may be viewed simultaneously in the display window.  All the
usual operations (zoom and pan, colortable enhancement, cursor readback, etc.)
still work for each frame even when in tile frames mode.
.TP 5
.B "-unix \fIname\fP"           
Specifies the unix domain socket name to use.  A "%d" in the filename will
be replaced with the user id.
.TP 5
.B "-unix_only"             
If set, only unix domain sockets will be used for communication with a client
program, inet sockets and fifos will be disabled.

.SH "APPLICATION RESOURCES"
XImtool is implemented as a client program which is responsible for loading
the frame buffers/colormaps, communicating with clients, etc, and a
user-modifiable GUI file written as a Tcl script which handles all the user
interface details.  The \fIclient resources\fP described below will be common
to any user-defined GUI, the \fIgui resources\fP may change depending on how
extensively the GUI has been modified by the user.  Each of these components
has its own set of resources, but to the user setting them is the same as
with any other application.  

\fIGterm\fP widget resources (i.e. those for the main image window or 
colorbar) may be set as either client or GUI resources.  See the
\fIxgterm(1)\fP man page for a complete description of \fIGterm\fP widget
resources.

.SS "CLIENT RESOURCES"
The client resources generally define the initial state of the application
or set configuration parameters.  
.RS
.TP 25
.B "Resource Name"
\fBDefault Value\fP
.sp -0.5
.TP 25
defConfig
1
.sp -0.5
.TP 25
defNFrames
0
.sp -0.5
.TP 25
tileBorderWidth
3
.sp -0.5
.TP 25
tileBorderColor
9
.sp -0.5
.TP 25
autoscale
false
.sp -0.5
.TP 25
antialias
false
.sp -0.5
.TP 25
antialiasType
boxcar
.sp -0.5
.TP 25
tileFrames
false
.sp -0.5
.TP 25
highlightFrames
true
.sp -0.5
.TP 25
gui
default
.sp -0.5
.TP 25
imtoolrc
/usr/local/lib/imtoolrc
.sp -0.5
.TP 25
invert
false
.sp -0.5
.TP 25
memModel
fast
.sp -0.5
.TP 25
basePixel:
64
.sp -0.5
.TP 25
maxColors:
216
.sp -0.5
.TP 25
cmapInitialize:
false
.sp -0.5
.TP 25
cmap1
none
.sp -0.5
.TP 25
cmap2
none
.sp -0.5
.TP 25
cmapDir1
none
.sp -0.5
.TP 25
cmapDir2
/usr/local/lib/imtoolcmap
.sp -0.5
.TP 25
input_fifo
/dev/imt1i
.sp -0.5
.TP 25
output_fifo
/dev/imt1o
.sp -0.5
.TP 25
unixaddr
/tmp/.IMT%d
.sp -0.5
.TP 25
port
5137
.sp -0.5
.TP 25
ism_addr
/tmp/.ISM%d
.sp -0.5
.TP 25
ism_task
"ism_wcspix.e wcspix &"
.RE

Description of ximtool client resources:

.TP 18
.B "defConfig"
Default frame buffer configuration number on startup.  See below for more
information on frame buffers.
.TP 18
.B "defNFrames"
Default number of frames on startup.  Set to zero to use the value from 
the frame buffer configuration (\fIimtoolrc\fP) file.  
.TP 18
.B "tileBorderWidth"
.sp -0.5
.TP 18
.B "tileBorderColor"
Used by the tile frames option.  Specifies how far
apart to space the frames in tile frames mode.
Color "9" refers to the Gterm widget resource color9,
which is assigned a color with its own resource.
.TP 18
.B "autoscale"
Enable/disable the autoscale option.
.TP 18
.B "antialias"
Enable/disable the antialias option.
.TP 18
.B "antialiasType"
Type of antialiasing.  Options include \fIboxcar\fP (default), \fIbilinear\fP,
\fInearest\fP, \fIarea\fP, \fIblkavg\fP, \fIlowpass\fP, and \fIgaussian\fP.
.TP 18
.B "tileFrames"
Enable/disable the tile frames option.
.TP 18
.B "highlightFrames"
Determines whether the current frame is highlighted when in tile frames mode.
.TP 18
.B "gui"
The GUI to be executed.  "default" refers to the default, builtin ximtool GUI.
You can replace this with your own GUI file if you are bold enough, and
completely change the look and functionality of the GUI if desired.
.TP 18
.B "imtoolrc"
Where to find the imtoolrc file.  This defines the
recognized frame buffer configurations.
.TP 18
.B "invert"
Start Ximtool using an inverted colormap.  When set, a "normalized" display
will always be the inverse of the selected colormap.
.TP 18
.B "memModel"
Determines how ximtool uses memory in the ximtool client and the X server.  
The options are \fIfast\fP, \fIbeNiceToServer\fP, and \fIsmall\fP.
The default is \fIfast\fP, which uses server pixmaps to make frame blink fast.
This is recommended unless server memory is very limited.  Note that even in
fast mode, the server pixmap is only the size of the display window, so
memory usage is reasonable even if the frame buffer is very large.
.sp -0.5
.TP 18
.B "basePixel"
.sp -0.5
.TP 18
.B "maxColors"
These two resources determine the region of colormap space used to
render image pixels.
.TP 18
.B "cmapInitialize"
Initialize the ximtool colormap at startup.  This is a required resource to
clear a previous ximtool colormap allowing a new \fBbasePixel\fP and
\fBmaxColors\fP to take effect.
.TP 18
.B "cmap1"
.sp -0.5
.TP 18
.B "cmap2"
User colormap files.  The intent here is to allow individual colormaps to be
conveniently specified as a resource.
.TP 18
.B "cmapDir1"
.sp -0.5
.TP 18
.B "cmapDir2"
User or system colormap directories.  By default cmapDir2 points to the system
directory /usr/local/lib/imtoolcmap, allowing a set of site default colormaps
to be defined here.  This leaves cmapDir1 available to a user colormap 
directory.
.TP 18
.B "input_fifo"
.sp -0.5
.TP 18
.B "output_fifo"
The input and output fifos for fifo i/o.  "Input" and "output" are from the
client's point of view.  Note that only one display server can use a
fifo-pair at one time.
.TP 18
.B "unixaddr"
Template address for unix domain socket.  The user must have write permission
on this directory, or the file must already exist.  \fI%d\fP, if given,
is replaced by the user's UID.
.TP 18
.B "port"
TCP/IP port for the server.  Note that only one server can listen on a port 
at one time, so if multiple ximtool servers are desired on the same
machine, they should be given different ports.
.TP 18
.B "ism_addr"
Template address for ISM unix domain socket. The user must have write
permission on this directory, or the file must already exist.  \fI%d\fP,
if given, is replaced by the user's UID.
.TP 18
.B "ism_task"
Command string to execute for the real-time pixel and WCS readout ISM

CHANGES.md
# X11IRAF important changes

## Version 2.1 (2021-07-17)

 * __IRAF community maintainance__

 * __64 bit compatibility__ (#2)

 * __External libraries removed__

   External libraries that were included as convenience copies were
   removed. The code is updated to use the current versions of these
   libraries, which are included in macOS as well as in the Linux
   distributions:
  
   - Tcl (#14),
   - xpm (#16),
   - xaw3d (#16)

   The unused CDL (#11) and vximtool (#18) subdirectories were also
   removed. They are maintained in a separate repository.

 * __Imake build system replaced__
 
   The Imake build system (xmkmf) was replaced by plain Makefiles
   (#19). Re-introduce support for macOS (#17).

 * __Several bugfixes and spellfixes__ (#12, #13, #21, #26, #27, #28). 
   Consequently use stdarg.h (#3)

 * __Clarify licenses and add licensing information__ (#20).
   Make it easy to remove non-free code.

 * __Remove unused code files__,
   and fix a number of compiler warnings. (#21, #34)

 * __Reorganize README and installation instructions__ (#22, #30, #36).

## Selected NOAO versions

This list is compiled from the available release information in the
original packages.

### Version 2.0BETA (2008-11-17)

 * 24-bit display Support for XImtool

 * 24-bit display Support for XGterm applications using imaging

 * General platform support improvements

### Version 1.5dev

### Version 1.3.1

### Version 1.3 (2002-05-06)

 *  real-time cursor image WCS and actual pixel readout

 *  new cursor centroiding function

 *  ability to "auto-register" frames

 *  integrated/enhanced control panel

 *  support for up to 16 display frames

 *  improved support for image tiling

 *  a "compass" indicator for the display orientation

 *  pixel table readout
 
 *  image header display (keyword selection, WCS display)
 
 *  horizontal/vertical cut-graphs of the display
 
 *  ruler markers on the display
 
 *  control port for external process communications

### Version 1.2 (2000-04-28)

### Version 1.1 (1997-09-08)

### Version 1.0 patch 1 (1997-04-05)

### Version 1.0 (1997-03-06)

 * initial version


ximtool/ximtool.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xmu/Error.h>
#include <Obm.h>

#define XIMTOOL_MAIN
#include "ximtool.h"

/*
 * XIMTOOL -- X11 based image display server and standalone image display
 * client.  This program is based on the IRAF prototype widget server (object
 * manager library or OBM) and the gterm-image widget.
 */

/* Data. */
XtAppContext app_context;
static char server[] = "server";

/* The builtin default GUI. */
char *defgui_text[] = {
#   include "ximtool.gui.h"
#ifndef OSI_COMPLIANT
    "send helpText addCallback anchorSelected anchor",
    "send helpText addCallback testAnchor testAnchor",
#endif
    NULL
};

void xim_onsig();
void Usage();
void printoption();


/* MAIN -- XImtool main program.  This is the only ximtool routine containing
 * window system specific code.
 */
int
main (argc, argv)
int argc;
char *argv[];
{
	register XimDataPtr xim = &ximtool_data;
	register int i;
        register IsmModule ism;

	Widget toplevel;
	XtPointer obm;
	char **sv_argv, *init_file = NULL, *str;
	int sv_argc, ncolors, base;
	int tile = 0;

	extern IsmModule ismNameToPtr();
	int 	xerror(), xioerror();


	/* Process the command line arguments.  Scan the arglist first to see
	 * if we're requesting help or a printout of the default GUI, if so 
	 * print these out and exit.
	 */
	if (argc > 1) {
	    if (strcmp (argv[1], "-help") == 0) {
      	        Usage ();
      	        exit (1);

    	    } else if (strcmp (argv[1], "-version") == 0) {
		printf ("Version:  %s\n", X11IRAF_VERSION);
      	        exit (1);

    	    } else if (strcmp (argv[1], "-defgui") == 0) {
	        register int i;
	        for (i=0;  defgui_text[i];  i++)
		    printf ("%s\n", defgui_text[i]);
	        exit (0);
	    }
	}


	/* Loop over the command line options and preprocess the ones that
	 * are widget/GUI resources we want to make available more easily.
	 * To do this we'll tweak the argument list so it appears to be a
 	 * "-xrm" resource setting, this means the X initialization code
	 * below will do all the real work.
	 */
	for (i=1; i < argc; i++) {

	    if (strcmp (argv[i], "-cmapName") == 0) {
		str = argv[++i];
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*cmapName:%s\0", str);

	    } else if (strcmp (argv[i], "-maxColors") == 0) {
		ncolors = atoi (argv[++i]);
		ncolors = max (32, min (201, ncolors));
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*maxColors:%d\0", ncolors);

	    } else if (strcmp (argv[i], "-basePixel") == 0) {
		base = atoi (argv[++i]);
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*basePixel:%d", base);

	    } else if (strcmp (argv[i], "-cmapInitialize") == 0) {
		str = argv[++i];
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*cmapInitialize:%s", str);

	    } else if (strcmp (argv[i], "-displayPanner") == 0) {
		str = argv[++i];
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*displayPanner:%s", str);

           } else if (strcmp (argv[i], "-displayMagnifier") == 0) {
                str = argv[++i];
                strcpy (argv[i-1], "-xrm\0");
                argv[i] = (char *) malloc (256);
                sprintf (argv[i], "XImtool*displayMagnifier:%s", str);

	    } else if (strcmp (argv[i], "-displayCoords") == 0) {
		str = argv[++i];
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*displayCoords:%s", str);

	    } else if (strcmp (argv[i], "-printConfig") == 0) {
		str = argv[++i];
		strcpy (argv[i-1], "-xrm\0");
		argv[i] = (char *) malloc (256);
		sprintf (argv[i], "XImtool*printConfig:%s", str);
	    }
	}

        /* Get local copy of argc and argv.  */
        if ((sv_argc = argc) > 0) {
            sv_argv = (char **) XtMalloc (argc * sizeof(char *));
 	    memmove (sv_argv, argv, argc * sizeof(char *)); 
        } else
            sv_argv = argv;

	/* Initialize applications context.  We don't use the top level
	 * shell created here, but we need to deal with it in order to
	 * get the application resources.  The object manager opens its own
	 * connection to the display server and manages a separate window
	 * hierarchy with its own top level shell.
	 */
	xim->toplevel = toplevel = XtAppInitialize (&app_context, "XImtool",
	    (XrmOptionDescList) NULL, 0, &sv_argc, sv_argv,
	    (String *) NULL, (ArgList) NULL, 0);

	/* Free saved arglist. */
	free ((char *)sv_argv);

	/* Get application resources. */
	XtVaGetApplicationResources (xim->toplevel, xim,
	    resources, XtNumber(resources),
	    /* Add any resource overrides here */
	    NULL);

	/* Check to see if we have the correct visual to proceed, if not
	 * shut down more gracefully and informatively than the BadMatch
	 * error from X that awaits us.
	 */

#ifdef PSEUDOCOLOR_ONLY
	screen = XtScreen (toplevel);
    	visual = DefaultVisualOfScreen(screen);
    	depth = DefaultDepthOfScreen(screen);
    	if (depth != 8 || visual->class != PseudoColor)
	    xim_badVisual (depth, visual->class);
#endif



	/* Initialize the object manager. */
	xim->obm = obm = (XtPointer) ObmOpen (app_context, argc, argv);
	ObmAddCallback (obm, OBMCB_setGterm|OBMCB_preserve, xim_reset,
	    (XtPointer)xim);


	/* Loop over the command line options. The default xim structure
	 * should be defined at this point so the command options can be
	 * used to override them.
	 */
	for (i=1; i < argc; i++) {

	    /* Anything without a '-' is a file name to load at startup.
	     * Only use the first name defined.
	     */
	    if (argv[i][0] != '-') { 			    /* File name */
		if (!init_file) {
		    init_file = argv[i];
		    if (access (init_file, F_OK) < 0) {
			fprintf (stderr, "%s: File does not exist: '%s'\n", 
			    argv[0], init_file);
			exit (1);
		    } else if (access (init_file, R_OK) < 0) {
			fprintf (stderr, 
			    "%s: File doesn't have read permission: '%s'\n", 
			    argv[0], init_file);
			exit (1);
		    }
		}

	    } else if (strcmp (argv[i], "-gui") == 0) {
		xim->gui = argv[++i];

	    } else if (strcmp (argv[i], "-cmap1") == 0) {
		xim->userCMap1 = argv[++i];

	    } else if (strcmp (argv[i], "-cmap2") == 0) {
		xim->userCMap2 = argv[++i];

	    } else if (strcmp (argv[i], "-cmapDir1") == 0) {
		xim->userCMapDir1 = argv[++i];

	    } else if (strcmp (argv[i], "-cmapDir2") == 0) {
		xim->userCMapDir2 = argv[++i];

	    } else if (strcmp (argv[i], "-imtoolrc") == 0) {
		xim->imtoolrc = argv[++i];

	    } else if (strcmp (argv[i], "-memModel") == 0) {
		xim->memModel = argv[++i];

	    } else if (strcmp (argv[i], "-config") == 0) {
		xim->def_config = atoi (argv[++i]);

	    } else if (strcmp (argv[i], "-nframes") == 0) {
		i++;
		xim->def_nframes = min (MAX_FRAMES, atoi (argv[i]));

	    } else if (strcmp (argv[i], "-tile") == 0) {
                tile = ++xim->tileFrames;

	    } else if (strcmp (argv[i], "-invert") == 0) {
		xim->invert++;

	    } else if (strcmp (argv[i], "-fifo") == 0) {
		if (strcmp ("none", argv[i+1]) == 0) {
		    xim->input_fifo = "none";
		} else {
                    xim->input_fifo = malloc (strlen(argv[++i])+2);
                    xim->output_fifo = malloc (strlen(argv[i])+2);
		    sprintf (xim->input_fifo, "%si", argv[i]);
		    sprintf (xim->output_fifo, "%so", argv[i]);
		}

	    } else if (strcmp (argv[i], "-port") == 0) {
		xim->port = atoi (argv[++i]);

	    } else if (strcmp (argv[i], "-nports") == 0) {
		xim->nports = atoi (argv[++i]);

	    } else if (strcmp (argv[i], "-unix") == 0) {
		xim->unixaddr = argv[++i];

	    } else if (strcmp (argv[i], "-fifo_only") == 0) {
		xim->unixaddr = "none";
		xim->port = 0;

	    } else if (strcmp (argv[i], "-inet_only") == 0 ||
	        strcmp (argv[i], "-port_only") == 0 ) {
		    xim->input_fifo = "";
		    xim->unixaddr = "none";

	    } else if (strcmp (argv[i], "-unix_only") == 0) {
		xim->input_fifo = "";
		xim->port = 0;

	    } else if (strcmp (argv[i], "-ismdev") == 0) {
		if (strcmp ("none", xim->ism_addr) )
		    xim->ism_addr = argv[++i];
		else 
		    i++;


	    /* Skip any standard X toolkit flags, they're handled above. 
	     */
	    } else if (strcmp (argv[i], "-bg") == 0) {
		i++;
	    } else if (strcmp (argv[i], "-fg") == 0) {
		i++;
	    } else if (strcmp (argv[i], "-iconic") == 0) {
		   ;
	    } else if (strcmp (argv[i], "-display") == 0) {
		i++;
	    } else if (strcmp (argv[i], "-geometry") == 0) {
		i++;
	    } else if (strcmp (argv[i], "-title") == 0) {
		i++;
	    } else if (strcmp (argv[i], "-xrm") == 0) {
		i++;

	    } else {
		fprintf (stderr, "Unrecognized flag '%s'\n", argv[i]);
      	        Usage();
      	        exit (1);
	    }
	}


	/* Initialize the ximtool/obm client code. */
	xim_clientOpen (xim);

	/* Load the Ximtool GUI.  If the GUI name is "default" the builtin
	 * default GUI is used.  This is stored as an array of text lines,
	 * which we must append newlines to and concatenate together to
	 * form the GUI message.
	 */
	if (strcmp (xim->gui, "default") == 0 ||
	    (ObmDeliverMsgFromFile (obm, server, xim->gui) != 0)) {

	    register char *ip, *op;
	    char *message;
	    int i;

	    message = (char *) malloc (409600);
	    for (i=0, op=message;  (ip = defgui_text[i]);  i++) {
		while (*ip)
		    *op++ = *ip++;
		*op++ = '\n';
	    }
	    *op++ = '\0';

	    ObmDeliverMsg (obm, server, message);
	    free ((char *)message);

	}

	/* Activate the GUI. */
	ObmActivate (obm);

	/* Initialize the frame buffers and graphics pipeline. */
	xim->tileFrames = tile;
	xim_initialize (xim, xim->def_config, xim->def_nframes, 1);

	/* Listen for a client connection. */
	xim_iisOpen (xim);
	xim_ismOpen (xim);

	/* Display a pretty logo. */
	xim_displayLogo (xim);

	/* Initialize the hardcopy option and printer configuration. */
	xim_initPrinterOps (xim);

	/* Open the file Load/Save structures. */
	xim_initLoad (xim);
	xim_initSave (xim);

	/* Load a file at startup if it was defined. */
	if ( init_file != NULL )
	    xim_loadFile (xim, init_file, 1);

        /* Lookup the ISM command for the WCSPIX task and start it.
         */
        if ((ism = ismNameToPtr ("wcspix"))) {
            system (ism->command);
	    ism_message (ism, "wcspix", "initialize");
	}

        XSetErrorHandler(xerror);
        XSetIOErrorHandler(xioerror);


/*	signal (SIGINT, (SIGFUNC)xim_onsig);*/

	/* EXECUTE */
	XtAppMainLoop (app_context);

	xim_shutdown (xim);
}


/* XIM_SHUTDOWN -- Terminate ximtool.
 */
int
xim_shutdown (xim)
register XimDataPtr xim;
{
	eps_close (xim->psim);
	xim_loadClose (xim);
	xim_saveClose (xim);
	xim_clientClose (xim);
	xim_iisClose (xim);
	xim_ismClose (xim);
	xim_close (xim);
	ObmClose (xim->obm);
	exit (0);
}


/* XIM_ADDINPUT -- Register a procedure to be called when there is input
 * to be processed on the given input source.  The ximtool code doesn't
 * talk to X directly so we need to provide this interface routine.
 */
XtInputId
xim_addInput (xim, input, proc, client_data)
register XimDataPtr xim;
int input;
void (*proc)();
XtPointer client_data;
{
	return XtAppAddInput (app_context, input,
	    (XtPointer)XtInputReadMask, *proc, client_data);
}


/* XIM_REMOVEINPUT -- Remove a callback previously posted with xim_addInput.
 */
void
xim_removeInput (xim, id)
register XimDataPtr xim;
XtPointer id;
{
	XtRemoveInput ((XtInputId)id);
}


/* USAGE -- Print a list of command-line options.
 */
void
Usage ()
{
    fprintf (stderr, "Usage:\n\n");
    printoption ("    ximtool");
    printoption ("[-basePixel <num>]");   	/* base cmap pixel        */
    printoption ("[-cmap1 <file>]");	   	/* User cmap 1            */
    printoption ("[-cmap2 <file>]");    	/* User cmap 2            */
    printoption ("[-cmapDir1 <dir>]");   	/* User cmapDir 1         */
    printoption ("[-cmapDir2 <dir>]");   	/* User cmapDir 2         */
    printoption ("[-cmapInitialize <bool>]");  	/* initialize colormap    */
    printoption ("[-cmapName <name>]");   	/* colormap name          */
    printoption ("[-config <num>]");	   	/* initial config         */
    printoption ("[-defgui]");		   	/* Print default GUI      */
    printoption ("[-displayPanner <bool>]");   	/* display panner box     */
    printoption ("[-displayMagnifier <bool>]");	/* display magnifier box  */
    printoption ("[-displayCoords <bool>]");   	/* display wcs coords box */
    printoption ("[-fifo <pipe>]");	    	/* fifo pipe              */
    printoption ("[-fifo_only]");  		/* use fifo only          */
    printoption ("[-gui <file>]");	    	/* GUI file               */
    printoption ("[-help]");		   	/* Print help             */
    printoption ("[-imtoolrc <file>]");   	/* fbconfig file          */
    printoption ("[-inet_only | -port_only]"); 	/* use inet only          */
    printoption ("[-invert]");       	   	/* invert colormap        */
    printoption ("[-ismdev <dev>]");      	/* ISM device template    */
    printoption ("[-maxColors <num>]");   	/* # of colors            */
    printoption ("[-memModel <type>]");   	/* memory model           */
    printoption ("[-nframes <num>]");    	/* # of frames            */
    printoption ("[-port <num>]");	    	/* inet port              */
    printoption ("[-printConfig <name>]");     	/* printer config file    */
    printoption ("[-tile]");       		/* tile frames            */
    printoption ("[-unix <name>]");	    	/* unix socket            */
    printoption ("[-unix_only]");  		/* use unix only          */
    printoption ("[<file>]");  		   	/* file to load           */
    fprintf (stderr,"\n");
}


/* PRINTOPTION -- Pretty-print an option string.
 */
static int cpos = 0;
void
printoption(st)
char 	*st;
{
  	if (strlen(st) + cpos > 78) {
    	    fprintf (stderr,"\n\t");
    	    cpos = 8;
  	}
  	fprintf (stderr,"%s ",st);
  	cpos = cpos + strlen(st) + 1;
}


/* xim_badVisual --  A bad X visual has been detected for the screen which
 * will cause us to crash with a BadMatch error.  Instead, abort with a
 * more informative message so the user can correct the visual.
 */
int
xim_badVisual (depth, class)
int	depth;
int	class;
{
    fprintf (stderr, 
	"\nERROR: Detected incorrect X visual:  depth=%d class=", depth);
        switch (class) {
        case StaticGray:    fprintf (stderr, "StaticGray\n")  ; break;
        case GrayScale:     fprintf (stderr, "GrayScale\n")   ; break;
        case StaticColor:   fprintf (stderr, "StaticColor\n") ; break;
        case PseudoColor:   fprintf (stderr, "PseudoColor\n") ; break;
        case TrueColor:     fprintf (stderr, "TrueColor\n")   ; break;
        case DirectColor:   fprintf (stderr, "DirectColor\n") ; break;
        default: 	    fprintf (stderr, "DirectColor\n") ; break;
        }

    fprintf (stderr,  "\n");
    fprintf (stderr, 
	"XImtool currently requires an 8-bit PseudoColor visual in order\n");
    fprintf (stderr, 
	"to operate properly. Unfortunately the only workaround for this\n");
    fprintf (stderr, 
	"at present is to start an 8-bit server using commands such as\n\n");
    fprintf (stderr, 
        "    %% startx -- -bpp 8\t\t\t\t\t    # XFree86 V3.x\n");
    fprintf (stderr, 
        "    %% startx -- -depth 8\t\t\t\t    # XFree86 V4.x\n");
    fprintf (stderr, 
        "    %% Xsun :0 -dev /dev/fb defclass PseudoColor defdepth 8  # Sun\n");
    fprintf (stderr, 
        "    %% Xdec -vclass0 PseudoColor\t\t\t\t    # Digital Unix\n");
    fprintf (stderr, 
	"\nThe actual commands used will vary depending on the platform,\n");
    fprintf (stderr, 
	"window manager/desktop used, and in some cases video hardware.\n");
    fprintf (stderr, 
	"See the Xserver(1) and xinit(1) man page for details. Users should\n\n");

    exit (1);
}



#define ERROR_XERROR    83      /* xerror: XError event */
#define ERROR_XIOERROR  84      /* xioerror: X I/O error */

/* XERROR -- Handle an XLIB server error.  A standard X error message is
 * printed and then the program either dumps core, exits, or ignores the error,
 * depending upon the value of the environment variable XGXERROR, if defined.
*/
/*ARGSUSED*/
int
xerror (display, event)
Display *display;
register XErrorEvent *event;
{
        static char *envvar = "XGXERROR";
        static int nerrs = 0;
        extern char *getenv();
        char fname[128];
        char *action;
        int pid;

        fprintf (stderr, "ximtool: warning, error event received:\n");
        (void) XmuPrintDefaultErrorMessage (display, event, stderr);
        if (nerrs++ > 50)
            exit (ERROR_XERROR);

        if ((action = getenv (envvar))) {
            if (strcmp (action, "dumpcore") == 0) {
                if ((pid = fork()) >= 0) {
                    if (pid) {
                        fprintf (stderr, "dumping core... ");
                        fflush (stderr);
                        sprintf (fname, "core.%d", pid);
                        wait(NULL); rename ("core", fname);
                        fprintf (stderr, "core file core.%d written\n", pid);
                       fflush (stderr);
                    } else
                        kill (getpid(), 6);
                } else
                    fprintf (stderr, "fork failed, no core dump produced\n");
           } else if (strcmp (action, "exit") == 0) {
                fprintf (stderr, "program terminated\n");
                exit (ERROR_XERROR);
            } else
                fprintf (stderr, "%s: unknown action %s\n", envvar, action);
        }
	fflush (stderr);

        return (0);
}


/*ARGSUSED*/
int
xioerror(dpy)
Display *dpy;
{
    char *SysErrorMsg();

    (void) fprintf (stderr,
        "ximtool: fatal IO error %d (%s) or KillClient on X server \"%s\"\r\n",
        errno, SysErrorMsg (errno),
        DisplayString (dpy));

    exit (ERROR_XIOERROR);
}

void xt_error(message)
    String message;
{
    (void) fprintf (stderr, "ximtool Xt error: %s\n", message);
    exit (1);
}


char *SysErrorMsg (n)
    int n;
{
    return((n >= 0) ? (char *)strerror(n) : "unknown error");
}


/* XIM_ONSIG -- Catch interrupt and shutdown gracefully.
 */
void
xim_onsig (sig, code, scp)
int     sig;                    /* signal which was trapped     */
int     *code;                  /* not used */
int     *scp;                   /* not used */
{
        xim_shutdown (&ximtool_data);
}

